src/NUnitFramework/framework/Assert.cs

```diff
@@ -360,19 +360,18 @@ private static void ReportFailure(ConstraintResult result, string message, param
 
         private static void ReportFailure(string message)
         {
-            // Failure is recorded in <assertion> element in all cases
-            TestExecutionContext.CurrentContext.CurrentResult.RecordAssertion(
-                AssertionStatus.Failed, message, GetStackTrace());
-
             // If we are outside any multiple assert block, then throw
             if (TestExecutionContext.CurrentContext.MultipleAssertLevel == 0)
                 throw new AssertionException(message);
+
+            // Otherwise, just record the failure in an <assertion> element
+            var result = TestExecutionContext.CurrentContext.CurrentResult;
```

Comment:
This looks good in the context of the issue, but we have seen a lot of NullReferenceExceptions in this method reported since the 3.6 release. Some have been fixed, but I was wondering if we can think of to make the experience better? I hate hiding the error, but I think it might have always been hidden. Maybe,

```C#
if (TestExecutionContext.CurrentContext == null) 
    throw new AssertionException(message);
```